nD (counting) and multiset_derangements (generating)

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

A function for counting derangements of a sequence and another for generating the derangement of elements of the sequence is added.

#### Other comments

There have been many comments about combinatorial functions in general and possibilities for enhancement of the generation of the function added here in #6662. This PR is intended as a place to now elicit comments for improvement from the wider community.


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
  * combinatorial numbers enumeration function for deranngements, nD, added that is similar to nP, nC and nT.
* utilities
  * iterables function for generating derangements of a sequence was added
<!-- END RELEASE NOTES -->

smichr committed:
nD and multiset_derangements

oscargus commented:
LGTM!

smichr commented:
There is some ongoing discussion on the issue page about this, so I am changing this to "draft".

smichr committed:
more special cases

smichr committed:
add random_derangement

When an attempt fails should one start over or work at getting success? It seems like the latter, otherwise those derangements that fail more often would likely be returned less often.

smichr committed:
do not recreate rv during attempts and restart

Restart whole process if placement of items leads to dead end.

smichr committed:
Update iterables.py

rathmann committed:
finish_derangement added

rathmann commented on sympy/utilities/iterables.py:
```diff
@@ -2285,14 +2285,244 @@ def generate_involutions(n):
+            yield p
+        return
+
+    # 2) aaabbb-like
```
I don't think case #2 is needed any more -- the finish_derangements code handles these cases efficiently.
For example:
```
assert(['b', 'a'] * 25 ==
           list(list(d) for d in
                multiset_derangements_mod2(['a', 'b'] * 25)) [0]
```
returns instantly.

rathmann commented:
Is this PR at a state where we are mostly discussing the algorithm, or are we ready to consider the API issues as well?

If this were merged now, wouldn't we have two competing multiset derangements routines?   (the old `generate_derangements()` and the new `multiset_derangements()`)

Near as I can tell, the new code is a great improvement over the old, **except** that there are two changes in semantics:

1. The old code emits derangements relative to a cannonical order, the new code is relative to the argument order.
2. The old gave the derangements in lexicographical order, the new produces a predictable-but-idiosyncratic order.

I suspect that both of these are fine, but should be mentioned in the documentation and release notes.

smichr committed:
remove redundant case

smichr commented:
> I suspect that both of these are fine, but should be mentioned in the documentation and release notes.

They both work the same way and can be used interchangeably. Since `generate_derangements` was there first, `multiset_derangements` can be thought of as a helper for it.

smichr committed:
accept int input to nD

smichr commented:
I wonder if `nD` should support a flag like `counts` or `multiplicity' which would allow `s` to be interpreted as multiplicities rather than elements so [1,1,1,2,2] or {1:3, 2:2} would correspond to 'abcddee'.

smichr commented:
I wonder if the special case of only 1 repeated element should be special cased (e.g. abcdefff).

smichr committed:
handle multiplicity and symbolic input

smichr committed:
keywords give input; no symbols

It seems the most natural to give input regarding the
set elements as opposed to the multiplicity of element
repetition (though I suppose it depends on who is using
the function). So specifying n distinct items or multiplicities
of element counts require use of keywords n and m to pass
that information, respectively.

nD('123') == nD(n=3) == nD(m=(1,1,1)) == nD(m={1:3}) == 2

smichr committed:
tests and doc edits

smichr commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -2175,3 +2177,124 @@ def eval(cls, n):
+    """
+    from sympy.core.compatibility import SYMPY_INTS
+    from sympy.utilities.iterables import multiset, multiset_derangements, iterable
+    from sympy.integrals.integrals import integrate, Integral
```
```suggestion
    from sympy.integrals.integrals import integrate
```

smichr commented on sympy/functions/combinatorial/tests/test_comb_numbers.py:
```diff
@@ -14,9 +14,7 @@
-
-
-x = Symbol('x')
+from sympy.abc import x, y
```
```suggestion
x = Symbol('x')
```

rathmann commented:
> I wonder if the special case of only 1 repeated element should be special cased (e.g. abcdefff).

Not sure how often this case arises, but if it is justified, I would think such a case might be wired in as 

```python
if mx ==1:   # Then ALL remaining elements to place are singletons
   # special case code - basically a set-based permutation algorithm
   . . .
```

That way any multiset which has a few singletons will get the special case treatment.  A generic permutation algorithm for the remaining elements, modified to respect the derangement constraints, should be reasonably easy to code in a way that is faster than having a stack of calls to subsets().

smichr commented on sympy/utilities/iterables.py:
```diff
@@ -2285,13 +2285,238 @@ def generate_involutions(n):
+        return [i for i in range(n) if rv[i] is None and s[i] != v]
+
+    def do(j):
+        if j == -1:
+            yield rv
```
```suggestion
        if j == 1:
            yield from finish_derangements()
```

smichr commented on sympy/utilities/iterables.py:
```diff
@@ -2168,13 +2168,238 @@ def generate_involutions(n):
+
+def generate_derangements(perm):
+    """
+    Routine to generate unique derangements or sets or multisets.
```
```suggestion
    Routine to generate unique derangements of sets or multisets.
```

smichr commented on sympy/utilities/iterables.py:
```diff
@@ -2195,9 +2420,21 @@ def generate_derangements(perm):
+    if not has_dups(perm):
+        s = perm
+        if len(perm) == 2:
+            yield [s[1],s[0]]
+            return
+        if len(perm) == 3:
+            yield [s[1],s[2],s[0]]
+            yield [s[2],s[0],s[1]]
```
```suggestion
            yield [s[1], s[0]]
            return
        if len(perm) == 3:
            yield [s[1], s[2], s[0]]
            yield [s[2], s[0], s[1]]
```

smichr commented:
@rathman, I don't have additions to this. Let's commit this since it is a significant improvement over what we already have.
